service/share/sample.go

```diff
@@ -0,0 +1,77 @@
+// TODO(@Wondertan): Instead of doing sampling over the coordinates do a random walk over NMT trees.
+package share
+
+import (
+	crand "crypto/rand"
+	"fmt"
+	"math/big"
+)
+
+// Sample is a point in 2D space over square.
+type Sample struct {
+	Row, Col int
+}
+
+// SampleSquare randomly picks *num* unique points from the given *width* square
+// and returns them as samples.
+func SampleSquare(squareWidth int, num int) ([]Sample, error) {
+	ss := newSquareSampler(squareWidth, num)
+	err := ss.generateSample(num)
+	if err != nil {
+		return nil, err
+	}
+	return ss.samples(), nil
+}
+
+type squareSampler struct {
+	squareWidth int
+	smpls       map[Sample]struct{}
+}
+
+func newSquareSampler(squareWidth int, expectedSamples int) *squareSampler {
+	return &squareSampler{
+		squareWidth: squareWidth,
+		smpls:       make(map[Sample]struct{}, expectedSamples),
+	}
+}
+
+// generateSample randomly picks unique point on a 2D spaces.
+func (ss *squareSampler) generateSample(num int) error {
+	if num > ss.squareWidth*ss.squareWidth {
+		return fmt.Errorf("number of samples %d exceeds square width %d", num, ss.squareWidth)
+	}
+
+	done := 0
+	for done < num {
+		s := Sample{
+			Row: randInt(ss.squareWidth),
+			Col: randInt(ss.squareWidth),
+		}
+
+		if _, ok := ss.smpls[s]; ok {
+			continue
+		}
+
+		done++
+		ss.smpls[s] = struct{}{}
+	}
+
+	return nil
+}
```

Comment:
If picking a large number of samples this could take a while. But a large number of samples would imply a small square size, so short runtime anyways. That being said, this could be a source of non-deterministic timeouts/performance. Would it be better to simply use a standard library function to generate `num` random numbers between `0` and `ss.squareWidth*ss.squareWidth`, then mutate them into 2D?

Reply:
I assume this is not a final solution and the next plan I see here is to do https://github.com/celestiaorg/celestia-node/issues/182, also as mentioned at the top of the file.

Reply:
Unfortunately, Golang std does not have such func in crypto safe randomization API https://pkg.go.dev/crypto/rand@go1.17.3.

Reply:
Also, note this code was working fine before for MVP.

Reply:
I understand your concern entirely and this __may__ loop forever, but this is not realistic for 16 samples we have as a default rn.

Reply:
@adlerjohn, can we keep this code as it is right now as it proved that it is functional and change it later to random tree walks as in #182?

Reply:
Absolutely we can keep it as-is for now, so long as there's a plan (which it looks like there is).